File: code/vl03_word_count.java
/*
 * Java version of VL03 EX0 using regex matching, thanks Marvin G.
 *
 * You probably need to rename this to main.java in order to compile and run.
 */

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.*;

public class Main {
	public static void main(String[] args) {
		ArrayList<String> doublequoted = new ArrayList<String>();
		ArrayList<String> unqoted = new ArrayList<String>();

		String Text = "apple orange banana \"honey pie\" sun \"high noon\"";
		Pattern p = Pattern.compile("[^\\s\"']+|\"([^\"]*) \"|'([^']*)'");
		Matcher m = p.matcher(Text);
		while(m.find()) {
			if(m.group(1) != null) {
				//Woerter mit Anfuehrungszeichen
				doublequoted.add(m.group(1));
			} else {
				//Woerter ohne Anfuehrungszeichen
				unqoted.add(m.group());
			}
		}

		System.out.println("all expressions: " + (doublequoted.size() + unqoted.size()));
		System.out.println("compound expression: " + doublequoted.size());
	}
}

File: code/vl03_word_count.pl
#!/usr/bin/perl
#
# Consider a string containing multiple space separated expressions. An
# expression is either a single word or a compound expression consisting of
# multiple words between double quotes.
#
# The (\w+|"[^"]+") regex is used here to match separate
# single or compound expressions.
#

# Example input
$input = q{apple orange banana "honey pie" sun "high noon"};

# Get an array of all expressions
@all_exp = $input =~ /(\w+|"[^"]+")/g;

# Get an array of compound expression, each begins with single "
@compound_exp = grep(/^"/, @all_exp);

print "all expressions: ", scalar @all_exp, "\n";
print "compound expressions: ", scalar @compound_exp, "\n";
